exec: LIKE broken for patterns which begin and end with wildcards

On master, if we add `5node-dist-vec` config to `distsql_numtables` logic test file and run it, we get:
```
SELECT str FROM NumToStr WHERE y < 10 AND str LIKE '%e%' ORDER BY y;
	-- FAIL
--- done: testdata/logic_test/distsql_numtables with config 5node-dist-vec: 11 tests, 1 failures

--- total: 0 tests, 0 failures
test logs left over in: /var/folders/j8/33cl3dhx5jn6j29mtlmxhfhm0000gn/T/logTestLogic707575344
--- FAIL: TestLogic (0.72s)
    test_log_scope.go:77: test logs captured to: /var/folders/j8/33cl3dhx5jn6j29mtlmxhfhm0000gn/T/logTestLogic707575344
    test_log_scope.go:58: use -show-logs to present logs inline
    --- FAIL: TestLogic/5node-dist-vec (0.70s)
        --- FAIL: TestLogic/5node-dist-vec/distsql_numtables (0.70s)
            logic.go:2338: 
                 
                testdata/logic_test/distsql_numtables:82: SELECT str FROM NumToStr WHERE y < 10 AND str LIKE '%e%' ORDER BY y
                expected:
                    one
                    three
                    five
                    seven
                    eight
                    nine
                    
                but found (query options: "") :
                    
                
            logic.go:2370: 
                testdata/logic_test/distsql_numtables:92: error while processing
            logic.go:2371: testdata/logic_test/distsql_numtables:92: too many errors encountered, skipping the rest of the input
FAIL
make: *** [testbaselogic] Error 1
```
I'm guessing the problem is with LIKE operator since previous queries without it have succeeded.

--- a/pkg/sql/exec/like_ops.go
+++ b/pkg/sql/exec/like_ops.go
@@ -45,33 +45,54 @@ func GetLikeOperator(
 		return NewNoop(input), nil
 	}
 	if len(pattern) > 1 && !strings.ContainsAny(pattern[1:len(pattern)-1], "_%") {
-		// Special cases for patterns which are just a prefix or suffix.
-		if pattern[0] == '%' {
+		// There are no wildcards in the middle of the string, so we only need to
+		// use a regular expression if both the first and last characters are
+		// wildcards.
+		firstChar := pattern[0]
+		lastChar := pattern[len(pattern)-1]
+		if !isWildcard(firstChar) && !isWildcard(lastChar) {
+			// No wildcards, so this is just an exact string match.
+			if negate {
+				return &selNEBytesBytesConstOp{
+					input:    input,
+					colIdx:   colIdx,
+					constArg: []byte(pattern),
+				}, nil
+			}
+			return &selEQBytesBytesConstOp{
+				input:    input,
+				colIdx:   colIdx,
+				constArg: []byte(pattern),
+			}, nil
+		}
+		if firstChar == '%' && !isWildcard(lastChar) {
+			suffix := []byte(pattern[1:])
 			if negate {
 				return &selNotSuffixBytesBytesConstOp{
 					input:    input,
 					colIdx:   colIdx,
-					constArg: []byte(pattern[1:]),
+					constArg: suffix,
 				}, nil
 			}
 			return &selSuffixBytesBytesConstOp{
 				input:    input,
 				colIdx:   colIdx,
-				constArg: []byte(pattern[1:]),
+				constArg: suffix,
 			}, nil
 		}
-		if pattern[len(pattern)-1] == '%' {
+		if lastChar == '%' && !isWildcard(firstChar) {
+			prefix := []byte(pattern[:len(pattern)-1])
 			if negate {
 				return &selNotPrefixBytesBytesConstOp{
 					input:    input,
 					colIdx:   colIdx,
-					constArg: []byte(pattern[:len(pattern)-1]),
+					constArg: prefix,
 				}, nil
 			}
 			return &selPrefixBytesBytesConstOp{
 				input:    input,
 				colIdx:   colIdx,
-				constArg: []byte(pattern[:len(pattern)-1]),
+				constArg: prefix,
 			}, nil
 		}
 	}
@@ -92,4 +113,8 @@ func GetLikeOperator(
 		colIdx:   colIdx,
 		constArg: re,
 	}, nil
+}
+
+func isWildcard(c byte) bool {
+	return c == '%' || c == '_'
 }

--- a/pkg/sql/exec/like_ops_test.go
+++ b/pkg/sql/exec/like_ops_test.go
@@ -16,46 +16,83 @@ import (
 	"regexp"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/exec/coldata"
 	"github.com/cockroachdb/cockroach/pkg/sql/exec/types"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 )
 
-func TestSelPrefixBytesBytesConstOp(t *testing.T) {
-	tups := tuples{{"abc"}, {"def"}, {"ghi"}}
-	runTests(t, []tuples{tups}, tuples{{"def"}}, orderedVerifier, []int{0}, func(input []Operator) (Operator, error) {
-		return &selPrefixBytesBytesConstOp{
-			input:    input[0],
-			colIdx:   0,
-			constArg: []byte("de"),
-		}, nil
-	})
-}
-
-func TestSelSuffixBytesBytesConstOp(t *testing.T) {
-	tups := tuples{{"abc"}, {"def"}, {"ghi"}}
-	runTests(t, []tuples{tups}, tuples{{"def"}}, orderedVerifier, []int{0}, func(input []Operator) (Operator, error) {
-		return &selSuffixBytesBytesConstOp{
-			input:    input[0],
-			colIdx:   0,
-			constArg: []byte("ef"),
-		}, nil
-	})
-}
-
-func TestSelRegexpBytesBytesConstOp(t *testing.T) {
-	tups := tuples{{"abc"}, {"def"}, {"ghi"}}
-	pattern, err := regexp.Compile(".e.")
-	if err != nil {
-		t.Fatal(err)
+func TestLikeOperators(t *testing.T) {
+	for _, tc := range []struct {
+		pattern  string
+		negate   bool
+		tups     tuples
+		expected tuples
+	}{
+		{
+			pattern:  "def",
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"def"}},
+		},
+		{
+			pattern:  "def",
+			negate:   true,
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"abc"}, {"ghi"}},
+		},
+		{
+			pattern:  "de%",
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"def"}},
+		},
+		{
+			pattern:  "de%",
+			negate:   true,
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"abc"}, {"ghi"}},
+		},
+		{
+			pattern:  "%ef",
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"def"}},
+		},
+		{
+			pattern:  "%ef",
+			negate:   true,
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"abc"}, {"ghi"}},
+		},
+		{
+			pattern:  "_e_",
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"def"}},
+		},
+		{
+			pattern:  "_e_",
+			negate:   true,
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"abc"}, {"ghi"}},
+		},
+		{
+			pattern:  "%e%",
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"def"}},
+		},
+		{
+			pattern:  "%e%",
+			negate:   true,
+			tups:     tuples{{"abc"}, {"def"}, {"ghi"}},
+			expected: tuples{{"abc"}, {"ghi"}},
+		},
+	} {
+		runTests(
+			t, []tuples{tc.tups}, tc.expected, orderedVerifier, []int{0},
+			func(input []Operator) (Operator, error) {
+				ctx := tree.MakeTestingEvalContext(cluster.MakeTestingClusterSettings())
+				return GetLikeOperator(&ctx, input[0], 0, tc.pattern, tc.negate)
+			})
 	}
-	runTests(t, []tuples{tups}, tuples{{"def"}}, orderedVerifier, []int{0}, func(input []Operator) (Operator, error) {
-		return &selRegexpBytesBytesConstOp{
-			input:    input[0],
-			colIdx:   0,
-			constArg: pattern,
-		}, nil
-	})
 }
 
 func BenchmarkLikeOps(b *testing.B) {

